File: packages/typegen/src/lib/transformer/syntax-type-checker/scopes/scope.ts
import type { ModuleScope } from "./module-scope";
import * as ts from "typescript";
import { ModuleIdentifier } from "rttist";

export enum InfoKind {
	NamedDeclaration,
	ImportDeclaration,
	TypeParameterDeclaration,
	/**
	 * Any declaration creating a type: interface, class, type alias etc.
	 */
	AnyTypeDeclaration,
}

export type ImportDeclarationInfo =
	| {
			kind: InfoKind.ImportDeclaration;
			namespaceImport: false;
			declaredName: string;
			moduleId: ModuleIdentifier;
			declaration: ts.ImportDeclaration;
	  }
	| {
			kind: InfoKind.ImportDeclaration;
			namespaceImport: true;
			moduleId: ModuleIdentifier;
			declaration: ts.ImportDeclaration;
	  };

export type NamedDeclarationInfo = {
	kind: InfoKind.NamedDeclaration;
	declaration: ts.NamedDeclaration;
};

export type TypeParameterDeclarationInfo = {
	kind: InfoKind.TypeParameterDeclaration;
	declaration: ts.TypeParameterDeclaration | ts.ClassDeclaration | ts.InterfaceDeclaration | ts.TypeAliasDeclaration;
};

export type AnyTypeDeclarationInfo = {
	kind: InfoKind.AnyTypeDeclaration;
	declaration: ts.ClassDeclaration | ts.InterfaceDeclaration | ts.TypeAliasDeclaration;
};

export type DeclarationInfo = NamedDeclarationInfo;
export type TypeDeclarationInfo = TypeParameterDeclarationInfo | AnyTypeDeclarationInfo;

export class Scope {
	protected readonly declarations = new Map<string, DeclarationInfo>();
	protected readonly typeDeclarations: Map<string, TypeDeclarationInfo>;

	/**
	 * It's not exactly module scope; it's the top scope that is tracked.
	 * It should be scope of the SourceFile (so module scope).
	 */
	public readonly moduleScope: ModuleScope;

	/**
	 *
	 * @param originator Node that created this scope.
	 * @param parent Parent scope.
	 */
	constructor(originator: ts.Node, parent: Scope);
	constructor(
		private readonly originator: ts.Node,
		protected readonly parent?: Scope
	) {
		this.typeDeclarations = hasTypeParameters(originator)
			? new Map(
					originator.typeParameters.map((tp) => [
						tp.name.getText(),
						{
							kind: InfoKind.TypeParameterDeclaration,
							declaration: tp,
						},
					])
			  )
			: new Map();

		this.moduleScope = (parent?.moduleScope ?? this) as ModuleScope;
	}

	addDeclaration(name: string, declaration: DeclarationInfo): void {
		this.declarations.set(name, declaration);
	}

	addTypeDeclaration(name: string, declaration: TypeDeclarationInfo): void {
		this.typeDeclarations.set(name, declaration);
	}

	/**
	 * Return type declaration by name.
	 * @param name
	 */
	getTypeDeclaration(name: string): TypeDeclarationInfo | ImportDeclarationInfo | undefined {
		return (
			this.typeDeclarations?.get(name) ||
			this.parent?.getTypeDeclaration(name) ||
			this.moduleScope.getImportDeclaration(name)
		);
	}

	/**
	 * Return declaration by name.
	 * @param name
	 */
	getDeclaration(name: string): NamedDeclarationInfo | ImportDeclarationInfo | undefined {
		if (name === "this") {
			return {
				kind: InfoKind.NamedDeclaration,
				declaration: this.getContextScope()?.originator as ts.NamedDeclaration,
			};
		}

		return (
			this.declarations.get(name) ||
			this.parent?.getDeclaration(name) ||
			this.moduleScope.getImportDeclaration(name)
		);
	}

	private getContextScope(): Scope | undefined {
		let scope: Scope | undefined = this;

		do {
			if (scope.isContextNode()) {
				return scope;
			}
			scope = scope.parent;
		} while (scope !== undefined);

		return undefined;
	}

	/**
	 * Check if the originator creates "this" context.
	 */
	private isContextNode(): boolean {
		return (
			ts.isClassLike(this.originator) ||
			ts.isFunctionDeclaration(this.originator) ||
			ts.isFunctionExpression(this.originator) ||
			ts.isObjectLiteralExpression(this.originator)
		);
	}
}

function hasTypeParameters(
	node: ts.Node
): node is ts.Node & { typeParameters: ts.NodeArray<ts.TypeParameterDeclaration> } {
	return (node as any).typeParameters !== undefined;
}

File: packages/plugins/rust/crates/rttist-swc-transformer/src/syntax_type_checker/scope.rs
use std::collections::HashMap;
use std::net::ToSocketAddrs;
use swc_core::ecma::ast;
use crate::syntax_type_checker::module_scope::ModuleScope;

// pub enum InfoKind {
//     NamedDeclaration,
//     ImportDeclaration,
//     TypeParameterDeclaration,
//     AnyTypeDeclaration,
// }
//
// pub struct DeclarationInfo;
// pub struct TypeParameterDeclarationInfo {
//
// }
// pub struct AnyTypeDeclarationInfo;
// pub struct TypeDeclarationInfo;
// // pub type TypeDeclarationInfo = TypeParameterDeclarationInfo + AnyTypeDeclarationInfo;
//
// pub struct Scope {
//     declarations: HashMap<String, DeclarationInfo>,
//     type_declarations: HashMap<String, TypeDeclarationInfo>,
//     module_scope: ModuleScope,
// }
// use std::collections::HashMap;
use crate::generate_module_id::ModuleIdentifier;

enum InfoKind {
    NamedDeclaration,
    ImportDeclaration,
    TypeParameterDeclaration,
    AnyTypeDeclaration,
}

pub struct ImportDeclarationInfo {
    kind: InfoKind,
    namespace_import: bool,
    declared_name: String,
    module_id: ModuleIdentifier,
    declaration: ImportDeclaration,
}

pub struct DeclarationInfo {
    kind: InfoKind,
    class_declaration: Option<ast::ClassDecl>,
    function_declaration: Option<ast::FnDecl>,
    variable_declaration: Option<ast::VarDecl>,
}

struct TypeDeclarationInfo {
    kind: InfoKind,
    type_parameter_declaration: Option<ast::TsTypeParam>,
    class_declaration: Option<ast::ClassDecl>,
    function_declaration: Option<ast::FnDecl>,
    interface_declaration: Option<ast::TsInterfaceDecl>,
    type_alias_declaration: Option<ast::TsTypeAliasDecl>,
}


struct Originator {
    class_declaration: Option<ast::ClassDecl>,
    function_declaration: Option<ast::FnDecl>,
    block_statement: Option<ast::BlockStmt>,
}

// trait HasTypeParameters {
//
// }
struct Scope {
    declarations: HashMap<String, DeclarationInfo>,
    type_declarations: HashMap<String, TypeDeclarationInfo>,
    module_scope: ModuleScope,
    originator: ast::Ident,
    parent: Option<Box<Scope>>,
}

impl Scope {
    fn new(
        originator: Originator,
        parent: Option<Box<Scope>>,
        type_parameters: Vec<ast::TsTypeParam>,
    ) -> Scope {
        let type_declarations = type_parameters
            .iter()
            .map(|tp| {
                (
                    tp.name.sym.to_string(),
                    TypeDeclarationInfo {
                        kind: InfoKind::TypeParameterDeclaration,
                        type_parameter_declaration: Some(tp.clone()),
                        class_declaration: None,
                        interface_declaration: None,
                        function_declaration: None,
                        type_alias_declaration: None,
                    }
                )
            })
            .collect();

        Scope {
            declarations: HashMap::new(),
            type_declarations,
            module_scope: parent.as_ref().map_or_else(
                || ModuleScope::new(),
                |p| p.module_scope.clone(),
            ),
            originator,
            parent,
        }
    }

    fn add_declaration(&mut self, name: String, declaration: DeclarationInfo) {
        self.declarations.insert(name, declaration);
    }

    fn add_type_declaration(&mut self, name: String, declaration: TypeDeclarationInfo) {
        self.type_declarations.insert(name, declaration);
    }

    fn get_type_declaration(&self, name: &str) -> Option<TypeDeclarationInfo> {
        self.type_declarations
            .get(name)
            .cloned()
            .or_else(|| self.parent.as_ref().and_then(|p| p.get_type_declaration(name)))
            .or_else(|| self.module_scope.get_import_declaration(name))
    }

    fn get_declaration(&self, name: &str) -> Option<NamedDeclarationInfo> {
        if name == "this" {
            return Some(NamedDeclarationInfo {
                kind: InfoKind::NamedDeclaration,
                declaration: self.get_context_scope()?.originator.clone(),
            });
        }

        self.declarations
            .get(name)
            .cloned()
            .or_else(|| self.parent.as_ref().and_then(|p| p.get_declaration(name)))
            .or_else(|| self.module_scope.get_import_declaration(name))
    }

    fn get_context_scope(&self) -> Option<&Scope> {
        let mut scope = self;

        while let Some(parent) = &scope.parent {
            if scope.is_context_node() {
                return Some(scope);
            }
            scope = parent;
        }

        None
    }

    fn is_context_node(&self) -> bool {
        matches!(
            self.originator,
            Node::ClassLike(_)
                | Node::FunctionDeclaration(_)
                | Node::FunctionExpression(_)
                | Node::ObjectLiteralExpression(_)
        )
    }
}

fn has_type_parameters(node: &Node) -> bool {
    node.type_parameters.is_some()
}

// struct ModuleScope {
//     // Define ModuleScope fields here
// }
//
// impl ModuleScope {
//     fn new() -> ModuleScope {
//         // Initialize ModuleScope fields here
//         ModuleScope {}
//     }
//
//     fn get_import_declaration(&self, name: &str) -> Option<ImportDeclarationInfo> {
//         // Implement getting import declaration here
//         None
//     }
// }

// struct ImportDeclaration {
//     // Define ImportDeclaration fields here
// }
//
// struct NamedDeclaration {
//     // Define NamedDeclaration fields here
// }
//
// struct TypeParameterDeclaration {
//     // Define TypeParameterDeclaration fields here
// }
//
// struct AnyTypeDeclaration {
//     // Define AnyTypeDeclaration fields here
// }